### DISCUSSION TITLE
How to restart an Uno Platform app?

### DISCUSSION THREAD

**Discussion by sciencekiller** [unoplatform/uno#15154] [Q&A]
Is there an API in Uno Platform that supports automatic application restart? If not, how do you implement this feature?

**Comment by limeniye** [comment]
Hello, @sciencekiller.
Currently [AppInstance](https://learn.microsoft.com/en-us/windows/apps/windows-app-sdk/applifecycle/applifecycle-restart#examples) and [CoreApplication.RequestRestartAsync](https://learn.microsoft.com/en-us/uwp/api/windows.applicationmodel.core.coreapplication.requestrestartasync?view=winrt-22621) is not implemented in Uno.

I can suggest you the following code which is support for WinUI, Android, Skia platforms:
(I guess it's also might be supported for Linux)

```C#
    private void Button_Click(object sender, RoutedEventArgs e)
    {
#if ANDROID
        var context = ContextHelper.Current;
        var packageManager = context.PackageManager;
        var intent = packageManager.GetLaunchIntentForPackage(context.PackageName);
        var componentName = intent.Component;

        var mainIntent = Android.Content.Intent.MakeRestartActivityTask(componentName);
        mainIntent.SetPackage(context.PackageName);
        context.StartActivity(mainIntent);
        Java.Lang.Runtime.GetRuntime().Exit(0);
#elif WINDOWS
        Microsoft.Windows.AppLifecycle.AppInstance.Restart("");
#elif HAS_UNO // Only for Skia. Use HAS_UNO_SKIA https://platform.uno/docs/articles/platform-specific-csharp.html
        var applicationPath = AppDomain.CurrentDomain.BaseDirectory;
        AppDomain.CurrentDomain.ProcessExit += (s, e) =>
        {
            System.Diagnostics.Process.Start(applicationPath + $"UnoApp53.Skia.Wpf.exe");
        };

        Application.Current.Exit();
#endif
    }
```

**Comment by sciencekiller** [comment]
Thanks for your advice. However, it didn't work well. In Uno, ```Microsoft.Windows.AppLifecycle``` didn't exist. It throw [CS0234](https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/compiler-messages/cs0234). By the way, I only need to run the app on Windows currently(I may add other platforms later, but I'm not sure) so you just need to give the code that support Windows.

**Reply by limeniye** [reply]
Please send the versions of your packages.
Or, check if you added the WINDOWS condition.

**Reply by limeniye** [reply]
The code of **HAS_UNO** block also will works for **WINDOWS**. 
But do not forget to change the name of the out .exe file:
```C#
#if WINDOWS
        //Microsoft.Windows.AppLifecycle.AppInstance.Restart("");

        var applicationPath = AppDomain.CurrentDomain.BaseDirectory;
        AppDomain.CurrentDomain.ProcessExit += (s, e) =>
        {
            Process.Start(applicationPath + $"UnoApp53.Windows.exe");
        };

        Application.Current.Exit();
#endif
```

**Reply by limeniye** [reply]
Or you can use the following code block
```C#
#if WINDOWS
AppRestartFailureReason restartError = Microsoft.Windows.AppLifecycle.AppInstance.Restart(null);
#endif
```
This 3 suggestions you can test on the following sample: 
[UnoAppRestartApplication.zip](https://github.com/unoplatform/uno/files/14022190/UnoAppRestartApplication.zip)
